src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -0,0 +1,598 @@
+            if (output != null && output.endsWith(synchronizationString)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Submit a task to start listening for output
+     */
+    private void startListeners() {
+        // Submit runnable task to start capturing.
+        if (stdOutFuture == null) {
+            stdOutFuture = executorService.submit(
+                    new OutputCapture(
+                            new CapturedStreamOutputSnapshot(settings.getStdoutSettings(), process.getInputStream(), System.out),
+                            ProcessStream.Stdout,
+                            this.getClass().getSimpleName(),
+                            controllerId));
+        }
+        if (!settings.isRedirectErrorStream() && stdErrFuture == null) {
+            // don't waste a callable on stderr if its just going to be redirected to stdout
+            stdErrFuture = executorService.submit(
+                    new OutputCapture(
+                            new CapturedStreamOutputSnapshot(settings.getStderrSettings(), process.getErrorStream(), System.err),
+                            ProcessStream.Stderr,
+                            this.getClass().getSimpleName(),
+                            controllerId));
+        }
+    }
+
+    /**
+     * Stops the process from running and tries to ensure the process is cleaned up properly.
+     * NOTE: sub-processes started by process may be zombied with their parents set to pid 1.
+     * NOTE: capture threads may block on read.
+     */
+    @Override
+    protected void tryDestroy() {
+        if (stdErrFuture != null && !stdErrFuture.isDone()) {
+            boolean isCancelled = stdErrFuture.cancel(true);
+            if (!isCancelled) {
+                logger.error("Failure cancelling stderr task");
+            }
+        }
+        if (stdOutFuture != null && !stdOutFuture.isDone()) {
+            boolean isCancelled = stdOutFuture.cancel(true);
+            if (!isCancelled) {
+                logger.error("Failure cancelling stdout task");
+            }
+        }
+        if (process != null) {
+            // terminate the app by closing the process' INPUT stream
+            IOUtils.closeQuietly(process.getOutputStream());
+        }
+    }
+
+    /**
+     * Close the input stream, close the FIFO, and wait for the remote process to terminate
+     * destroying it if necessary.
+     *
+     * @return true if the process was successfully destroyed
+     */
+    public boolean terminate() {
+        closeFIFO();
+        tryDestroy();
+        boolean exited = false;
+        try {
+            exited = process.waitFor(10, TimeUnit.SECONDS);
+            processJournal.close();
+            if (!exited) {
+                // we timed out waiting for the process to exit; it may be in a blocking call, so just force
+                // it to shutdown
+                process.destroy();
+                process.waitFor();
```

Comment:
Won't this wait forever if destroy failed somehow?  Can destroy fail?  It doesn't seem like it can.

Reply:
I think its correct; this should probably use the same timeout as above. Done.